Fix include gstreamer path on cross compiler toolchains

psaavedra opened the pull request:
#### 25efd1478b48b5406fdab3b7b9c1f7e0ef95a7ed
<pre>
Fix include gstreamer path on cross compiler toolchains
<a href="https://bugs.webkit.org/show_bug.cgi?id=241483">https://bugs.webkit.org/show_bug.cgi?id=241483</a>

Reviewed by Adrian Perez de Castro.

Set the include paths for the gstreamer components to the full path
using the find_path(). This function relies in CMAKE_FIND_ROOT_PATH to
find the right place where the includes they are. This fixes possible
warnings/errors on cross toolchains using -Wpoison-system-directories
and -Werror=poison-system-directories.

* Source/cmake/FindGStreamer.cmake:

Canonical link: <a href="https://commits.webkit.org/251895@main">https://commits.webkit.org/251895@main</a>
</pre>

aperezdc commented on Source/cmake/FindGStreamer.cmake:
```diff
@@ -77,8 +77,9 @@ macro(FIND_GSTREAMER_COMPONENT _component_prefix _pkgconfig_name _library)
     string(REGEX MATCH "(.*)/gstreamer-1.0" _dummy "${_include_dir}")
-    if ("${CMAKE_MATCH_1}" STREQUAL "")
-        set(${_component_prefix}_INCLUDE_DIRS "${_include_dir}/gstreamer-1.0;${PC_${_component_prefix}_INCLUDE_DIRS}")
+    find_path(_full_path_include_dir NAMES "${_include_dir}/gstreamer-1.0")
```
The call to `find_path()` should be inside the `if` block, as it is unneeded otherwise. Also `find_path()` will store the result variable in the cache, so second and subsequent calls to the `FIND_GSTREAMER_COMPONENT()` macro will *not* actually try to find anything because the variable will be already present in the cache—one needs an unique result variable name here (or using CMake 3.21+ which added a `NO_CACHE` flag; but this is not an option right now). Lastly, you don't need to manually check for the `-NOTFOUND` suffix, CMake already does that automatically inside `if` expressions.

I think what you want is something like this:

```cmake
# Populate the list initially from the _INCLUDE_DIRS result variable.
set(${_component_prefix}_INCLUDE_DIRS ${PC_${_component_prefix}_INCLUDE_DIRS})
set(_include_dir "${PC_${_component_prefix}_INCLUDEDIR}")
string(REGEX MATCH "(.*)/gstreamer-1.0" _dummy "${_include_dir}")
if ("${CMAKE_MATCH_1}" STREQUAL "")
    find_path(${_component_prefix}_RESOLVED_INCLUDEDIR NAMES "${_include_dir}/gstreamer-1.0")
    # Only add the resolved path from `_INCLUDEDIR` if found.
    if (${_component_prefix}_RESOLVED_INCLUDEDIR)
        list(APPEND ${_component_prefix}_INCLUDE_DIRS
             "${_component_prefix}_RESOLVED_INCLUDEDIR")
    endif ()
endif ()

aperezdc reviewed (changes requested):
Ideally we would rewrite the whole `FindGStreamer.cmake` file to use imported targets. For an example that uses imported targets _and_ deals with find-module components a good source of inspiration would be `FindGTK.cmake`.

Now, I understand that you may want to keep things as they are in general, and for now only fixing the compilation warnings when cross-compiling, so it's fine by to leave such rewrite to be done later.

Please take a look at my comment below, I think the issue with the cache can be the reason why the patch does not quite work as intended.

aperezdc commented on Source/cmake/FindGStreamer.cmake:
```diff
@@ -75,12 +75,20 @@ macro(FIND_GSTREAMER_COMPONENT _component_prefix _pkgconfig_name _library)
+        # Only add the resolved path from `_INCLUDEDIR` if found.
+        if (${_component_prefix}_RESOLVED_INCLUDEDIR)
+            list(APPEND ${_component_prefix}_INCLUDE_DIRS
+                 "${_component_prefix}_RESOLVED_INCLUDEDIR")
```
You want to use `${${_component_prefix}_RESOLVED_INCLUDEDIR}` during expansion, to resolve the additional level of indirection. This is not needed inside the `if()` command because it does that automatically by itself. CMake, huh 🤷🏼?

```suggestion
                 "${${_component_prefix}_RESOLVED_INCLUDEDIR}")
```

aperezdc replied:
Actually, I notice my suggestion from the previous review round was already wrong, sorry about that!

psaavedra committed:
Fix include gstreamer path on cross compiler toolchains
https://bugs.webkit.org/show_bug.cgi?id=241483

Reviewed by Adrian Perez de Castro.

Set the include paths for the gstreamer components to the full path
using the find_path(). This function relies in CMAKE_FIND_ROOT_PATH to
find the right place where the includes they are. This fixes possible
warnings/errors on cross toolchains using -Wpoison-system-directories
and -Werror=poison-system-directories.

* Source/cmake/FindGStreamer.cmake:

Canonical link: https://commits.webkit.org/251895@main
